### ISSUE TITLE
Ouput ansi-colors is not working

### ISSUE THREAD

**Issue by pati-ni** [nnicandro/emacs-jupyter#376]
Recently coloring stopped working when I execute source blocks in org-mode.

This is a minimal example:

```org-mode
#+PROPERTY: header-args:python :async yes :session thesession :exports both :results raw drawer


#+begin_src python
print('Test')
tea = {}
tea['the']
#+end_src

#+RESULTS:
:results:
: Test
# [goto error]
: [0;31m[0m
: [0;31mKeyError[0mTraceback (most recent call last)
: [0;32m<ipython-input-2-87a0654f2360>[0m in [0;36m<module>[0;34m[0m
: [1;32m      1[0m [0mprint[0m[0;34m([0m[0;34m'Test'[0m[0;34m)[0m[0;34m[0m[0;34m[0m[0m
: [1;32m      2[0m [0mtea[0m [0;34m=[0m [0;34m{[0m[0;34m}[0m[0;34m[0m[0;34m[0m[0m
: [0;32m----> 3[0;31m [0mtea[0m[0;34m[[0m[0;34m'the'[0m[0;34m][0m[0;34m[0m[0;34m[0m[0m
: [0m
: [0;31mKeyError[0m: 'the'
:end:

```

There are gibberish characters in the outputs that should be rendered as colors but org ignores those.

```
emacs --version
GNU Emacs 29.0.50
Copyright (C) 2021 Free Software Foundation, Inc.
```

**Comment by timlod** [comment]
Have a check if this solves your issue:
https://github.com/nnicandro/emacs-jupyter/issues/366#issuecomment-985730376

**Comment by timlod** [comment]
I removed `org-superstar` - the workaround mentioned I haven't tried. I posted an issue on this with `org-superstar` but it hasn't yet received any attention.

**Comment by pati-ni** [comment]
I do not have org-superstar enabled in my setup. I have hide leading stars enabled.

However, I was able to reproduce it independently with `emacs -q` and a pure python source block. I had just loaded `emacs-jupyter` but was outside a jupyter block.

**Comment by timlod** [comment]
Just to clarify, you can reproduce the issue with `emacs -q` and just org-mode and emacs-jupyter, even when you don't use `jupyter-python`? So the problem is not related to this package, but occurs in org when `org-hide-leading-stars`? If so, that could point to a common issue that occurs when using `org-superstar`.

**Comment by pati-ni** [comment]
> Just to clarify, you can reproduce the issue with `emacs -q` and just org-mode and emacs-jupyter, even when you don't use `jupyter-python`? So the problem is not related to this package, but occurs in org when `org-hide-leading-stars`? If so, that could point to a common issue that occurs when using `org-superstar`

I think I can reproduce it independently of `org-superstar` or `org-hide-leading-stars`.

-------------

Here are the steps:

1. Launch Emacs
`emacs -q`

2. Load a very minimal configuration

```elisp

(require 'package)

(package-initialize)
(require 'use-package)

(org-babel-do-load-languages
 'org-babel-load-languages
 '((emacs-lisp . t)
   (R . t)
   (latex . t)
   (python . t)
   (shell . t)
   ))
```

3. Evaluate a source block

![minifridge-2022-01-12_14:59:39-emacs-447x154](https://user-images.githubusercontent.com/14965608/149166144-8a9faf05-57e1-42fd-b3b6-fe7ff7861b44.png)

4. Evaluating your suggestion

![minifridge-2022-01-12_15:00:33-emacs-647x146](https://user-images.githubusercontent.com/14965608/149166256-f6677f17-c6e2-49d7-83ac-319eea318e69.png)

I get a direct conversion, without re executing the snippet, characters are rendered properly:

![minifridge-2022-01-12_15:00:12-emacs-388x158](https://user-images.githubusercontent.com/14965608/149166205-99fcd6c1-a1f0-4329-83c5-6fea6d59a48a.png)

5. Kill org buffer and revisit

![minifridge-2022-01-12_15:09:17-emacs-387x154](https://user-images.githubusercontent.com/14965608/149166764-399ed8cc-4235-48bb-8a96-8c70e38ef711.png)


All in all does not look related to the emacs-jupyter package.

I can spot 2 problems:

1. Colors are not rendered real-time
2. `ansi-color-apply-on-region` colors persist across buffer sessions and are lost if we kill the buffer and reopen

**Comment by timlod** [comment]
This is strange, I remember running pretty much the same when checking my in my own issue, and colours rendered just fine for me - unless I loaded `org-superstar`. I may verify this again on the new versions one of these days, this may be a separate regression?

**Comment by pati-ni** [comment]
I tried with an earlier version of emacs. I get the same behaviour unfortunately and I found it  confusing :S

It looks like `emacs-jupyter` had a workaround which did not require the use of the `org-babel-after-execute-hook` to apply the ansi-colors. Or my memory plays tricks on me.

@timlod what is the version of your working setup (emacs, org-mode, ansi-color and emacs-jupyter-version)? I suppose you do not have to apply ansi-colors through `org-babel-after-execute-hook`, right?

**Comment by timlod** [comment]
I had used Emacs 27.1, org 9.5.1. I had run this on Dec 3, so whatever Jupyter was available then.
In any case, with jupyter it also works fine on my full config which has this now:
- `GNU Emacs 27.1 (build 1, x86_64-pc-linux-gnu, X toolkit, Xaw3d scroll bars) of 2021-03-28, modified by Debian`
- `jupyter 20211130.1647`
- `Org mode version 9.5.2 (9.5.2-gfbff08)`

(both jupyter and vanilla python)

**Comment by akirakyle** [comment]
The solution I posted in #380 should work when kernels are asynchronous since it's not adding a hook but overriding the function `jupyter-ansi-color-apply-on-region` which uses the private function `ansi-color--find-face` which has since been removed from emacs.

**Comment by pati-ni** [comment]
@akirakyle oh yeah your solution is definitely more proper and elegant for the newer versions of emacs in which `ansi-color--find-face` is missing! I did not read the 380 thread in detail, i just identified the same symptom and my understanding of elisp code is limited to come to the same resolution on my own. Maybe you make a PR that branches the functionality according to the emacs version and push upstream?

**Comment by cvanelteren** [comment]
The fix in #380 does not work for me, is there another fix I can try?
